flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -526,62 +526,105 @@ class ExprLowering {
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->AIMAG()}}});
-      }
-      llvm_unreachable("array of complex unhandled");
-    } else if constexpr (TC == Fortran::lower::CharacterCat) {
-      return genCharLit<KIND>(con.GetScalarValue().value(), con.LEN());
+      using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
+      return genval(Fortran::evaluate::ComplexConstructor<KIND>{
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.REAL()}},
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.AIMAG()}}});
     } else {
       llvm_unreachable("unhandled constant");
     }
   }
 
+  template <Fortran::common::TypeCategory TC, int KIND>
+  mlir::Value genArrayLit(
+      const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
+          &con) {
+    // TODO:
+    // - Multi-dimensional array
+    using T = Fortran::evaluate::Type<TC, KIND>;
+    if (con.Rank() == 1) {
+      std::vector<Fortran::evaluate::Scalar<T>> values;
+      auto idxTy = builder.getIndexType();
+      fir::InsertValueOp res;
+      std::size_t counter{0};
+      if constexpr ((TC == Fortran::lower::IntegerCat) ||
+                    (TC == Fortran::lower::RealCat) ||
+                    (TC == Fortran::lower::ComplexCat) ||
+                    (TC == Fortran::lower::LogicalCat)) {
+        values = con.GetArrayValue();
+      } else {
+        llvm_unreachable("unhandled constant");
+      }
+      fir::SequenceType::Shape shape{(int64_t)(values.size())};
+      mlir::Type arrayTy =
+          fir::SequenceType::get(shape, converter.genType(TC, KIND));
+      auto array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
+      for (const Fortran::evaluate::Scalar<T> &value : values) {
+        auto idx = builder.createIntegerConstant(idxTy, counter);
+        auto constant = genScalar<TC, KIND>(value);
+        res = builder.create<fir::InsertValueOp>(
+            getLoc(), arrayTy,
+            (counter++ == 0) ? array.getResult() : res.getResult(), constant,
+            idx);
+      }
+      return res.getResult();
+    } else if (con.Rank() != 0) {
+      TODO(); // "Multi dimensional array not yet supported"
+    }
+    return mlir::Value{};
+  }
+
+  template <Fortran::common::TypeCategory TC, int KIND>
+  mlir::Value
+  genval(const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
+             &con) {
+    // TODO:
+    // - derived type constant
+    if (con.Rank() > 0) {
+      return genArrayLit(con);
+    } else {
+      using T = Fortran::evaluate::Type<TC, KIND>;
+      const std::optional<Fortran::evaluate::Scalar<T>> &opt =
+          con.GetScalarValue();
+      if (opt.has_value()) {
+        if constexpr (TC == Fortran::lower::CharacterCat)
+          return genCharLit<KIND>(opt.value(), con.LEN());
+        else
+          return genScalar<TC, KIND>(opt.value());
```

Comment:
With `if constexpr` it makes sense to add the braces, I think. It helps highlight that the `else` is more like a `#ELSE` than a pedestrian `if`-`else`.